tests/test_browser.go

```diff
@@ -0,0 +1,297 @@
+	// launch the browser
+	bt := chromium.NewBrowserType(ctx).(*chromium.BrowserType)
+	b := bt.Launch(rt.ToValue(launchOpts))
+	t.Cleanup(func() {
+		select {
+		case <-ctx.Done():
+		default:
+			b.Close()
+		}
+	})
+
+	tb := &browser{
+		ctx:     bt.Ctx, // This context has the additional wrapping of common.WithLaunchOptions
+		rt:      rt,
+		state:   state,
+		http:    testServer,
+		samples: samples,
+		Browser: b,
+	}
+	if enableFileServer {
+		tb = tb.withFileServer()
+	}
+	return tb
+}
+
+// withHandle adds the given handler to the HTTP test server and makes it
+// accessible with the given pattern.
+func (b *browser) withHandle(pattern string, handler http.HandlerFunc) *browser {
+	if b.http == nil {
+		panic("You should enable HTTP test server, see: withHTTPServer option")
+	}
+	b.http.Mux.Handle(pattern, handler)
+	return b
+}
+
+const testBrowserStaticDir = "static"
+
+// withFileServer serves a file server using the HTTP test server that is
+// accessible via `testBrowserStaticDir` prefix.
+//
+// This method is for enabling the static file server after starting a test
+// browser. For early starting the file server see withFileServer function.
+func (b *browser) withFileServer() *browser {
+	const (
+		slash = string(os.PathSeparator)
+		path  = slash + testBrowserStaticDir + slash
+	)
+
+	fs := http.FileServer(http.Dir(testBrowserStaticDir))
+
+	return b.withHandle(path, http.StripPrefix(path, fs).ServeHTTP)
+}
+
+// URL returns the listening HTTP test server's URL combined with the given path.
+func (b *browser) URL(path string) string {
+	if b.http == nil {
+		panic("You should enable HTTP test server, see: withHTTPServer option")
+	}
+	return b.http.ServerHTTP.URL + path
+}
+
+// staticURL is a helper for URL("/`testBrowserStaticDir`/"+ path).
+func (b *browser) staticURL(path string) string {
+	return b.URL("/" + testBrowserStaticDir + "/" + path)
+}
+
+// attachFrame attaches the frame to the page and returns it.
+func (b *browser) attachFrame(page api.Page, frameID string, url string) api.Frame {
+	pageFn := `
+	async (frameId, url) => {
+		const frame = document.createElement('iframe');
+		frame.src = url;
+		frame.id = frameId;
+		document.body.appendChild(frame);
+		await new Promise(x => frame.onload = x);
+		return frame;
+	}
+	`
+	return page.EvaluateHandle(
+		b.rt.ToValue(pageFn),
+		b.rt.ToValue(frameID),
+		b.rt.ToValue(url)).
+		AsElement().
+		ContentFrame()
+}
+
+// detachFrame detaches the frame from the page.
+// func (b *browser) detachFrame(page api.Page, frameID string) {
+// 	pageFn := `
+// 	frameId => {
+//         	document.getElementById(frameId).remove();
+//     	}
+// 	`
+// 	page.Evaluate(
+// 		b.rt.ToValue(pageFn),
+// 		b.rt.ToValue(frameID))
+// }
+
+// launchOptions provides a way to customize browser type
+// launch options in tests.
+type launchOptions struct {
+	Debug    bool   `js:"debug"`
+	Headless bool   `js:"headless"`
+	SlowMo   string `js:"slowMo"`
+	Timeout  string `js:"timeout"`
+}
```

Comment:
Could we reuse the existing [`common.LaunchOptions`](https://github.com/grafana/xk6-browser/blob/3aca99a846a08fc320880911696bd3d75ed123fa/common/launch.go#L40)? Otherwise we'd have to keep this in sync.

Reply:
I tried doing that but I couldn't do so because of some goja. When we pass `*common.LaunchOptions` to goja, it doesn't read `slowMo`. I discovered that it's because of field name mapping. Maybe you can help me figure out how?

Reply:
> have a separate launch function with native Go types that basically replicates what we do in Launch() 

This is the best way. I think we need to isolate goja-talking things with a facade from internal stuff. So we can better test and do stuff like that.